deps/redis/librdb-hidden-api.h

```diff
@@ -0,0 +1,27 @@
+/*
+ * The core library of LIBRDB is librdb.so. It is also responsible for linking
+ * and combining Redis object files from this folder. The extension library of
+ * LIBRDB (librdb-ext.so) also requires certain functionality from this folder,
+ * such as the use of the rax data structure or CRC.
+ *
+ * Instead of refactoring and creating an additional shared library, or making
+ * copies for each library (with all the associated implications) we shall expose
+ * specific symbols from librdb.so that are needed by librdb-ext.so. These symbols
+ * won't be documented in the API, but they will be available for use. Since we
+ * compile by default with hidden visibility, we need to mark them explicitly as
+ * visible using the following macro.
+ *
+ * While it might initially seem cumbersome to manually designate each function as
+ * visible, this approach has its benefits. By carefully selecting the functions
+ * to expose, we are encouraged to contemplate what we are exposing and why.
+ * Moreover, it helps prevent the cluttering clients of librdb with unnecessary
+ * symbols.
```

Comment:
@yossigo maybe you can think of a better idea or have a strong opinion here.
it means that whoever links with our library gets many (unprefixed) APIs exposed, and possibly colliding.
e.g. raxNext, d2string, etc.

Reply:
Another simple and effective option is to prefix the few functions with something like `RDB_HIDDEN_`.

Reply:
Is there a specific reason for making extension api a separate library? I mean, we can generate a single so file and still provide two headers. Users who want to use extension API can include extension header and use those extra functions. (As as bonus, maintaining single file is a lot easier than maintaining two files)

Reply:
We started with the easy choice of single library, but then raised aspects of modularity, seperation of concerns, maybe versioning, etc. And there is also my internal thought of having it in the far future as part of redis.

Reply:
To workaround the symbols issues and API exposure I already modified the code to have in the core librdb API:
```
typedef struct RdbNonDocApi RdbNonDocApi; /* incomplete struct */
_LIBRDB_API RdbNonDocApi *RDB_getNonDocApi();
```
If you don't mind maybe let's leave it as is, until it will meet us again. Or do it later with lower priority.

Reply:
that's being creative :)
another simple option is to just include these files (e.g. rax.c) in both projects, if the symbols are hidden, won't cause any issues other than a few more kb in the binaries.

Reply:
i'm assuming that they don't interact.
i.e. that we don't pass a rax pointer created by one lib, to be accessed by the other.
also, make sure that other than `visibility=hidden` you also have `-Bsymolic`

Reply:
They don't. 

Why to use `-Bsymbolic` and defer symbol resolution until runtime?

Reply:
AFAIK what it means is that local symbols are resolved within the library and won't attempt to be linked outside the library at runtime.
i.e. if your library uses raxSeek, even if its visibility is hidden, and external libraries won't be able to use it, there's still a chance your library will use a raxSeek provided by another library, instead of the local one

maybe i'm wrong. do you have other info?

Reply:
I misunderstood this flag based on obsolete documentation 
> This option is somewhat historic, in that it was primarily designed for use in building the runtime linker itself.

Now I read that 
> Alert linker to search this shared object before the executable for symbols

But all the internal functions got resolved and doesn't appear in the symbols of the shared object. Verified by `nm -D`. And if it weren't resolved ... can this flag might impact someone that is integrating our libraries.

Reply:
a library can use `malloc` and `free` and then when it is linked with an executable, it can override these and provide an alternative implementation.
i.e. even if both your libraries don't expose `raxNext` and they have their own implementation, they may end up using a `raxNext` that's provided by the executable.
the great risk is that in this case, they're both different versions of redis's rax, but with some small differences, and one can end up using inline functions from one rax.h and non-inline functions from another, and if some struct has a different layout it can be catastrophic. (based on real life experience with sds.c)

Reply:
>  even if both your libraries don't expose raxNext and they have their own implementation, they may end up using a raxNext that's provided by the executable.

I don't think it is true. If raxNext() doesn't appear in the dynamic symbol table (output of nm -D) of the library, it suggests that the symbol is not exposed for dynamic linking, which means it cannot be overridden in the usual dynamic linking manner. And therefore we don't have a problem and `-Bsymbolic` can be discarded.